### ISSUE TITLE
ValueError: Dimension 3 in both shapes must be equal

### ISSUE THREAD

**Issue by db-vis** [aydao/stylegan-convert-architecture#1]
I'm trying to convert my taki0112 model by using your stylegan-convert-architecture framework.
My model is progressively trained with a starting resolution of 8 to 512 (8, 16, 32, 64, 128, 256, 512). 

taki0112_to_nvlabs.py is able to load successfully my checkpoints but than it fails with the following error when building the  TensorFlow graph.

Here is the entire error message ->

 [*] Reading checkpoints...
 [*] Success to read StyleGAN.model-2405312
Constructing networks...

G                             Params    OutputShape         WeightShape
---                           ---       ---                 ---
latents_in                    -         (?, 512)            -
labels_in                     -         (?, 0)              -
lod                           -         ()                  -
dlatent_avg                   -         (512,)              -
G_mapping/latents_in          -         (?, 512)            -
G_mapping/labels_in           -         (?, 0)              -
G_mapping/PixelNorm           -         (?, 512)            -
G_mapping/Dense0              262656    (?, 512)            (512, 512)
G_mapping/Dense1              262656    (?, 512)            (512, 512)
G_mapping/Dense2              262656    (?, 512)            (512, 512)
G_mapping/Dense3              262656    (?, 512)            (512, 512)
G_mapping/Dense4              262656    (?, 512)            (512, 512)
G_mapping/Dense5              262656    (?, 512)            (512, 512)
G_mapping/Dense6              262656    (?, 512)            (512, 512)
G_mapping/Dense7              262656    (?, 512)            (512, 512)
G_mapping/Broadcast           -         (?, 16, 512)        -
G_mapping/dlatents_out        -         (?, 16, 512)        -
Truncation                    -         (?, 16, 512)        -
G_synthesis/dlatents_in       -         (?, 16, 512)        -
G_synthesis/4x4/Const         534528    (?, 512, 4, 4)      (512,)
G_synthesis/4x4/Conv          2885632   (?, 512, 4, 4)      (3, 3, 512, 512)
G_synthesis/ToRGB_lod7        1539      (?, 3, 4, 4)        (1, 1, 512, 3)
G_synthesis/8x8/Conv0_up      2885632   (?, 512, 8, 8)      (3, 3, 512, 512)
G_synthesis/8x8/Conv1         2885632   (?, 512, 8, 8)      (3, 3, 512, 512)
G_synthesis/ToRGB_lod6        1539      (?, 3, 8, 8)        (1, 1, 512, 3)
G_synthesis/Upscale2D         -         (?, 3, 8, 8)        -
G_synthesis/Grow_lod6         -         (?, 3, 8, 8)        -
G_synthesis/16x16/Conv0_up    2885632   (?, 512, 16, 16)    (3, 3, 512, 512)
G_synthesis/16x16/Conv1       2885632   (?, 512, 16, 16)    (3, 3, 512, 512)
G_synthesis/ToRGB_lod5        1539      (?, 3, 16, 16)      (1, 1, 512, 3)
G_synthesis/Upscale2D_1       -         (?, 3, 16, 16)      -
G_synthesis/Grow_lod5         -         (?, 3, 16, 16)      -
G_synthesis/32x32/Conv0_up    2885632   (?, 512, 32, 32)    (3, 3, 512, 512)
G_synthesis/32x32/Conv1       2885632   (?, 512, 32, 32)    (3, 3, 512, 512)
G_synthesis/ToRGB_lod4        1539      (?, 3, 32, 32)      (1, 1, 512, 3)
G_synthesis/Upscale2D_2       -         (?, 3, 32, 32)      -
G_synthesis/Grow_lod4         -         (?, 3, 32, 32)      -
G_synthesis/64x64/Conv0_up    1442816   (?, 256, 64, 64)    (3, 3, 512, 256)
G_synthesis/64x64/Conv1       852992    (?, 256, 64, 64)    (3, 3, 256, 256)
G_synthesis/ToRGB_lod3        771       (?, 3, 64, 64)      (1, 1, 256, 3)
G_synthesis/Upscale2D_3       -         (?, 3, 64, 64)      -
G_synthesis/Grow_lod3         -         (?, 3, 64, 64)      -
G_synthesis/128x128/Conv0_up  426496    (?, 128, 128, 128)  (3, 3, 256, 128)
G_synthesis/128x128/Conv1     279040    (?, 128, 128, 128)  (3, 3, 128, 128)
G_synthesis/ToRGB_lod2        387       (?, 3, 128, 128)    (1, 1, 128, 3)
G_synthesis/Upscale2D_4       -         (?, 3, 128, 128)    -
G_synthesis/Grow_lod2         -         (?, 3, 128, 128)    -
G_synthesis/256x256/Conv0_up  139520    (?, 64, 256, 256)   (3, 3, 128, 64)
G_synthesis/256x256/Conv1     102656    (?, 64, 256, 256)   (3, 3, 64, 64)
G_synthesis/ToRGB_lod1        195       (?, 3, 256, 256)    (1, 1, 64, 3)
G_synthesis/Upscale2D_5       -         (?, 3, 256, 256)    -
G_synthesis/Grow_lod1         -         (?, 3, 256, 256)    -
G_synthesis/512x512/Conv0_up  51328     (?, 32, 512, 512)   (3, 3, 64, 32)
G_synthesis/512x512/Conv1     42112     (?, 32, 512, 512)   (3, 3, 32, 32)
G_synthesis/ToRGB_lod0        99        (?, 3, 512, 512)    (1, 1, 32, 3)
G_synthesis/Upscale2D_6       -         (?, 3, 512, 512)    -
G_synthesis/Grow_lod0         -         (?, 3, 512, 512)    -
G_synthesis/images_out        -         (?, 3, 512, 512)    -
G_synthesis/lod               -         ()                  -
G_synthesis/noise0            -         (1, 1, 4, 4)        -
G_synthesis/noise1            -         (1, 1, 4, 4)        -
G_synthesis/noise2            -         (1, 1, 8, 8)        -
G_synthesis/noise3            -         (1, 1, 8, 8)        -
G_synthesis/noise4            -         (1, 1, 16, 16)      -
G_synthesis/noise5            -         (1, 1, 16, 16)      -
G_synthesis/noise6            -         (1, 1, 32, 32)      -
G_synthesis/noise7            -         (1, 1, 32, 32)      -
G_synthesis/noise8            -         (1, 1, 64, 64)      -
G_synthesis/noise9            -         (1, 1, 64, 64)      -
G_synthesis/noise10           -         (1, 1, 128, 128)    -
G_synthesis/noise11           -         (1, 1, 128, 128)    -
G_synthesis/noise12           -         (1, 1, 256, 256)    -
G_synthesis/noise13           -         (1, 1, 256, 256)    -
G_synthesis/noise14           -         (1, 1, 512, 512)    -
G_synthesis/noise15           -         (1, 1, 512, 512)    -
images_out                    -         (?, 3, 512, 512)    -
---                           ---       ---                 ---
Total                         26179768

D                    Params    OutputShape         WeightShape
---                  ---       ---                 ---
images_in            -         (?, 3, 512, 512)    -
labels_in            -         (?, 0)              -
lod                  -         ()                  -
FromRGB_lod0         128       (?, 32, 512, 512)   (1, 1, 3, 32)
512x512/Conv0        9248      (?, 32, 512, 512)   (3, 3, 32, 32)
512x512/Conv1_down   18496     (?, 64, 256, 256)   (3, 3, 32, 64)
Downscale2D          -         (?, 3, 256, 256)    -
FromRGB_lod1         256       (?, 64, 256, 256)   (1, 1, 3, 64)
Grow_lod0            -         (?, 64, 256, 256)   -
256x256/Conv0        36928     (?, 64, 256, 256)   (3, 3, 64, 64)
256x256/Conv1_down   73856     (?, 128, 128, 128)  (3, 3, 64, 128)
Downscale2D_1        -         (?, 3, 128, 128)    -
FromRGB_lod2         512       (?, 128, 128, 128)  (1, 1, 3, 128)
Grow_lod1            -         (?, 128, 128, 128)  -
128x128/Conv0        147584    (?, 128, 128, 128)  (3, 3, 128, 128)
128x128/Conv1_down   295168    (?, 256, 64, 64)    (3, 3, 128, 256)
Downscale2D_2        -         (?, 3, 64, 64)      -
FromRGB_lod3         1024      (?, 256, 64, 64)    (1, 1, 3, 256)
Grow_lod2            -         (?, 256, 64, 64)    -
64x64/Conv0          590080    (?, 256, 64, 64)    (3, 3, 256, 256)
64x64/Conv1_down     1180160   (?, 512, 32, 32)    (3, 3, 256, 512)
Downscale2D_3        -         (?, 3, 32, 32)      -
FromRGB_lod4         2048      (?, 512, 32, 32)    (1, 1, 3, 512)
Grow_lod3            -         (?, 512, 32, 32)    -
32x32/Conv0          2359808   (?, 512, 32, 32)    (3, 3, 512, 512)
32x32/Conv1_down     2359808   (?, 512, 16, 16)    (3, 3, 512, 512)
Downscale2D_4        -         (?, 3, 16, 16)      -
FromRGB_lod5         2048      (?, 512, 16, 16)    (1, 1, 3, 512)
Grow_lod4            -         (?, 512, 16, 16)    -
16x16/Conv0          2359808   (?, 512, 16, 16)    (3, 3, 512, 512)
16x16/Conv1_down     2359808   (?, 512, 8, 8)      (3, 3, 512, 512)
Downscale2D_5        -         (?, 3, 8, 8)        -
FromRGB_lod6         2048      (?, 512, 8, 8)      (1, 1, 3, 512)
Grow_lod5            -         (?, 512, 8, 8)      -
8x8/Conv0            2359808   (?, 512, 8, 8)      (3, 3, 512, 512)
8x8/Conv1_down       2359808   (?, 512, 4, 4)      (3, 3, 512, 512)
Downscale2D_6        -         (?, 3, 4, 4)        -
FromRGB_lod7         2048      (?, 512, 4, 4)      (1, 1, 3, 512)
Grow_lod6            -         (?, 512, 4, 4)      -
4x4/MinibatchStddev  -         (?, 513, 4, 4)      -
4x4/Conv             2364416   (?, 512, 4, 4)      (3, 3, 513, 512)
4x4/Dense0           4194816   (?, 512)            (8192, 512)
4x4/Dense1           513       (?, 1)              (512, 1)
scores_out           -         (?, 1)              -
---                  ---       ---                 ---
Total                23080225

Building TensorFlow graph...
Traceback (most recent call last):
  File "C:\Users\bd\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\framework\ops.py", line 1628, in _create_c_op
    c_op = c_api.TF_FinishOperation(op_desc)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Dimension 3 in both shapes must be equal, but are 32 and 512. Shapes are [1,1,3,32] and [1,1,3,512]. for 'Assign_7' (op: 'Assign') with input shapes: [1,1,3,32], [1,1,3,512].

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "taki0112_to_nvlabs.py", line 329, in <module>
    main()
  File "taki0112_to_nvlabs.py", line 264, in main
    update_weight = [tf.assign(new, old)]
  File "C:\Users\bd\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\ops\state_ops.py", line 221, in assign
    validate_shape=validate_shape)
  File "C:\Users\bd\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\ops\gen_state_ops.py", line 61, in assign
    use_locking=use_locking, name=name)
  File "C:\Users\bd\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\framework\op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "C:\Users\bd\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\util\deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "C:\Users\bd\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\framework\ops.py", line 3274, in create_op
    op_def=op_def)
  File "C:\Users\bd\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\framework\ops.py", line 1792, in __init__
    control_input_ops)
  File "C:\Users\bd\AppData\Local\Continuum\anaconda3\lib\site-packages\tensorflow\python\framework\ops.py", line 1631, in _create_c_op
    raise ValueError(str(e))
ValueError: Dimension 3 in both shapes must be equal, but are 32 and 512. Shapes are [1,1,3,32] and [1,1,3,512]. for 'Assign_7' (op: 'Assign') with input shapes: [1,1,3,32], [1,1,3,512].

**Comment by aydao** [comment]
I am checking this out now. I admittedly tested this script less than the other one, and I did suspect there could be an issue for progressive growing if you start with taki0112. My use case has been to do nvlabs -> taki0112 -> nvlabs and so on as needed. 

You are getting the error because the layers are not the same size. The reason for that likely has to do with them being misaligned: in the script I make an assumption that the layers are in the same order in both taki0112 and nvlabs architectures. Typically, that is true. I will try to track down when that is not the case and update the repo with a fixed script.

**Comment by aydao** [comment]
I have a potential fix, but I'd need you (or someone with a taki0112 model trained from scratch) to test it out. Check out [this new script](https://github.com/aydao/stylegan-convert-architecture/blob/progressive_fix/taki0112_reshape_progressive.py) on the "progressive-fix" branch.

**Make a backup of your model checkpoint, stored elsewhere, before attempting to run it.**

You can run `python taki0112_reshape_progresive.py --dataset [your data set] --kimg ###### --start_res 8 --img_size 512 --progressive True` and make sure that it drops a checkpoint with the counter set to 2 (StyleGAN.model-2). As a sanity check, make sure that the `checkpoint` file lists this new model, e.g., its contents would be:
```
model_checkpoint_path: "StyleGAN.model-2"
all_model_checkpoint_paths: "StyleGAN.model-2"
```
At that point you should be able to run the taki0112_to_nvlabs.py script as usual, and it should work, something like `python taki0112_to_nvlabs.py --dataset [your data set] --kimg ###### --start_res 8 --img_size 512 --progressive True`.

What the new script does is essentially pre-processing. It will copy over only the layers expected by taki0112_to_nvlabs.py and put this in a temporary checkpoint. The rest of the instructions should work as before. I've done a cursory run through and got it working for an untrained/noisy model. Hopefully it works as is. If not, I'll need someone to provide a taki0112 model trained from scratch to truly debug it and make sure the correct layers are being copied over.

**Comment by db-vis** [comment]
Hi Aydao, I've tried your proposed fix, but the problem still persists.
You can find in the link below a trained taki0112 network.

https://drive.google.com/drive/folders/1x3CwXseXHEHXU_boLRRTlxAhIf2Agcb3?usp=sharing

**Comment by aydao** [comment]
Thanks for sharing (cool model!), that really helped me figure out the problem. I got it convert. 

I believe the issue was with how my script was handling / expecting the starting resolution. In brief, I used the same process described above, but before running the taki0112_to_nvlabs.py script I had to change part of the directory's name from "8to512" to "512to512" and then use the flag set to "--start_res 512". I shall push a commit with my fixes to the branch soon, likely do some more testing before merging it to master, and close the issue then. Let me know if you have any issues. I'm also happy to delete my local copy of your models and remove the link above if you want to keep it private.
